File: 12-hash/6-common-patterns/main.py
# Common Hashing Problems - Python Example

from collections import Counter

# Anagram check
def are_anagrams(s1, s2):
    return Counter(s1) == Counter(s2)

print("Are 'listen' and 'silent' anagrams?", are_anagrams("listen", "silent"))

# Intersection of arrays
def array_intersection(arr1, arr2):
    set1 = set(arr1)
    return [x for x in arr2 if x in set1]

print("Intersection:", array_intersection([1,2,2,1], [2,2]))

# Two-sum problem
def two_sum(nums, target):
    hash_map = {}
    for i, num in enumerate(nums):
        complement = target - num
        if complement in hash_map:
            return (hash_map[complement], i)
        hash_map[num] = i
    return None

print("Two sum indices:", two_sum([2,7,11,15], 9))

File: 12-hash/6-common-patterns/main.java
// Common Hashing Problems - Java Example

import java.util.HashMap;
import java.util.HashSet;
import java.util.ArrayList;
import java.util.List;

public class CommonHashProblems {

    // Anagram check
    public static boolean areAnagrams(String s1, String s2) {
        if (s1.length() != s2.length()) return false;
        int[] freq = new int[256];  // Assuming ASCII
        for (char c : s1.toCharArray()) freq[c]++;
        for (char c : s2.toCharArray()) {
            freq[c]--;
            if (freq[c] < 0) return false;
        }
        return true;
    }

    // Intersection of arrays
    public static List<Integer> arrayIntersection(int[] arr1, int[] arr2) {
        HashSet<Integer> set1 = new HashSet<>();
        for (int num : arr1) set1.add(num);
        List<Integer> result = new ArrayList<>();
        for (int num : arr2) {
            if (set1.contains(num)) {
                result.add(num);
            }
        }
        return result;
    }

    // Two sum
    public static int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> hashMap = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (hashMap.containsKey(complement)) {
                return new int[]{hashMap.get(complement), i};
            }
            hashMap.put(nums[i], i);
        }
        return new int[]{-1, -1};
    }

    public static void main(String[] args) {
        System.out.println("Are 'listen' and 'silent' anagrams? " + areAnagrams("listen", "silent"));

        List<Integer> intersection = arrayIntersection(new int[]{1,2,2,1}, new int[]{2,2});
        System.out.print("Intersection: ");
        for (int x : intersection) {
            System.out.print(x + " ");
        }
        System.out.println();

        int[] indices = twoSum(new int[]{2,7,11,15}, 9);
        System.out.println("Two sum indices: (" + indices[0] + ", " + indices + ")");
    }
}
